director/profile.go

```diff
@@ -723,115 +738,174 @@ func VerifyMDMProfiles(profileListData types.ProfileListData, device types.Devic
-		// for i := range profilesToInstall {
-		// 	deviceProfile := profilesToInstall[i]
-		// 	if savedSharedProfile.PayloadIdentifier == deviceProfile.PayloadIdentifier {
-		// 		InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: savedSharedProfile.HashedPayloadUUID, ProfileIdentifier: savedSharedProfile.PayloadIdentifier, Message: "Shared Profile is a device profile, skipping"})
-		// 		found = true
-		// 		continue
-		// 	}
-		// }
+			// Profile is present, but the hash doesn't match
+			if profileForVerification.Installed && !hashMatches {
+				InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: profileForVerification.HashedPayloadUUID, ProfileIdentifier: profileForVerification.PayloadIdentifier, Message: "VerifyMDMProfiles: Profile is present but hashed payload UUID does not match", Metric: profileForVerification.Type})
+				sharedProfilesToInstall, profilesToInstall = addProfileToInstallLists(profileForVerification, sharedProfilesToInstall, profilesToInstall)
+			}
 
-		if !found {
-			InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: savedSharedProfile.HashedPayloadUUID, ProfileIdentifier: savedSharedProfile.PayloadIdentifier, Message: "Shared Profile is not installed"})
-			sharedProfilesToInstall = append(sharedProfilesToInstall, savedSharedProfile)
+			// Profile is present and hash matches, success
+			if profileForVerification.Installed && hashMatches {
+				InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: profileForVerification.HashedPayloadUUID, ProfileIdentifier: profileForVerification.PayloadIdentifier, Message: "VerifyMDMProfiles: Profile is present and hashed payload UUID matches", Metric: profileForVerification.Type})
+			}
+		} else { // Profile is not in the profileList
+			// But it should be installed
+			if profileForVerification.Installed {
+				InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: profileForVerification.HashedPayloadUUID, ProfileIdentifier: profileForVerification.PayloadIdentifier, Message: "VerifyMDMProfiles: Profile is present not in the profile list and should be installed", Metric: profileForVerification.Type})
+				sharedProfilesToInstall, profilesToInstall = addProfileToInstallLists(profileForVerification, sharedProfilesToInstall, profilesToInstall)
+			} else { // Not present, and shouldn't be installed
+				InfoLogger(LogHolder{DeviceUDID: device.UDID, DeviceSerial: device.SerialNumber, ProfileUUID: profileForVerification.HashedPayloadUUID, ProfileIdentifier: profileForVerification.PayloadIdentifier, Message: "VerifyMDMProfiles: Profile is not present and should not be installed", Metric: profileForVerification.Type})
+			}
 		}
 	}
 
+	devices = append(devices, device)
+	_, err = PushProfiles(devices, profilesToInstall)
+	if err != nil {
+		ErrorLogger(LogHolder{Message: err.Error()})
```

Comment:
Why do you log this error instead of return? In previous cases of errors, you wrap the error and return.

Reply:
Because an error shouldn't halt pushing / removing the other kinds of profiles

Reply:
Could WarnLogger or similar be used?